Feature/package actions

dscottboggs opened the pull request:
Route definitions for pkg actions

dscottboggs committed:
make prefix macro available to all routes

dscottboggs committed:
add package list route

dscottboggs committed:
Add route which cleans unused built packages

moved to separate method to keep ameba from taking a long time

also because it looked bad to have it so deeply nested

j8r commented on src/utils.cr:
```diff
@@ -20,6 +22,15 @@ macro fmt_route(route = "", namespace = false)
+# This macro only works inside of a route block with the variable "context"
+# available
+macro prefix
+  context.params.query["prefix"]? || DPPM.default_prefix
```
A macro for this?!  They shouldn't be abused (slower compile time, harder to debug). Use a method taking a context as argument.

j8r replied:
It can also be moved out of utils, for example in `Actions` or in `DppmRestApi`

j8r commented on src/actions/pkg.cr:
```diff
@@ -2,26 +2,114 @@ module DppmRestApi::Actions::Pkg
   ALL_PKGS = ""
   ONE_PKG  = "/:id"
+
+  def output_config_for(package, to builder)
```
Do like in the stdlib https://crystal-lang.org/api/master/JSON/Serializable.html#to_json%28json%3AJSON%3A%3ABuilder%29-instance-method for the typing. (`json : JSON::Builder`)
I suggest a better self-explanatory method name, what does this method do?
I guess `build_package_vars`.

dscottboggs replied:
> what does this method do?

It *outputs* the *config for package* into *builder*. I don't know how it could be more clear than that. Although, I agree more type restrictions are better.

dscottboggs replied:
I guess `build_config_response_for`? I suppose *output* isn't the right word.

j8r replied:
Actually this method won't build a config response _for_ a package, but from, and for the client. I think better to drop this `_for` suffix.

dscottboggs replied:
Ok.

j8r commented on src/actions/pkg.cr:
```diff
@@ -2,26 +2,114 @@ module DppmRestApi::Actions::Pkg
+    builder.field package.package do
+      builder.object do
+        package.pkg_file.config_vars.try &.each_key do |key|
+          builder.field key, package.get_config key
```
Use https://github.com/DFabric/dppm/blob/master/src/prefix/app.cr#L240 instead, else you'll miss library conf files like `/conf/php-fpm.ini`

dscottboggs replied:
I don't understand, this is for the package. The application configurations are a different route. Or am I misunderstanding how the app/pkg structure works?

j8r replied:
Yes sorry, I meant using `Pkg#each_config_key`, else dependencies configuration will be missed (php-fpm, in the future JVM).

dscottboggs committed:
Move reusable code into separate function

dscottboggs committed:
Add documentation, flush IOs

dscottboggs committed:
Add package delete route

dscottboggs committed:
Add package build route

Reorganize pkg code and start on testing

updated utils macros

Clean up and rebase

j8r commented on src/actions/pkg.cr:
```diff
@@ -2,49 +2,172 @@ module DppmRestApi::Actions::Pkg
+                # This method call  ^^ raises an untyped error if the key is not found
+              },
+              "errors" => [] of Nil,
+            }.to_json ctx.response
```
`JSON.build` can be used

j8r commented on src/actions/pkg.cr:
```diff
@@ -2,49 +2,172 @@ module DppmRestApi::Actions::Pkg
+            # catch the exception thown by the lack of a config value for the specified key
+            # otherwise, pass it on to the
+            msg = e.message || raise e
+            raise e unless msg.starts_with? "config key not found"
```
No need to do this. `get_config` has a block variant that yields when the key isn't found https://github.com/DFabric/dppm/blob/master/src/prefix/app.cr#L197
I'm creating a custom error anyway. Not sure what option is the best. On one hand using an exception brings a standard message, on the other hand the block one will be faster and more elegant.

j8r replied:
You can also now catch `ConfigKeyError`

j8r commented on src/utils.cr:
```diff
@@ -1,8 +1,9 @@
-  context
+require "dppm/prefix"
+
+def deny_access!(to ctx)
```
Its place is in the `Actions` module.

dscottboggs commented:
Ok that's what I was thinking. Thanks.

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,44 +1,146 @@
+                builder.field "dependencies" do
+                  deps.each do |name, version|
+                    semver = SemanticVersion.parse version
+                    DPPM::Prefix.new(prefix).each_pkg do |pkg|
```
Why iterating over all packages?

j8r replied:
"libraries" or "installed libraries" is a more correct word, because there are optional dependencies like `git` (for the filebrowser application).
You can use `Pkg#libs`.

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,44 +1,146 @@
+  end
+
+  def find_package_by_name(name, pfx) : DPPM::Prefix::Pkg?
+    DPPM::Prefix.new(pfx).each_pkg { |pkg| return pkg if pkg.package == name }
```
Instantiate directly the package, then checking if it exists.
Remember that a package name is `mypackage_1.0.0`.
Else, if we want info for `mypackage`, that's the source one (`Src`)

dscottboggs committed:
revise package actions

j8r commented on src/ext/dppm_pkg.cr:
```diff
@@ -0,0 +1,55 @@
+          builder.field key do
+            builder.string value: get_config key
+          rescue e
+            raise e unless e.starts_with? "config key not found"
```
Please use `ConfigKeyError`

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,45 +1,137 @@
+        # was found and we can query it
+        if key = context.params.query["get"]?
+          data = selected_pkg.get_config? key
+          if data.nil?
```
Use `ConfigKeyError`.
A config key can be present, with a nil value.

dscottboggs replied:
oof good catch, thank you

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,45 +1,137 @@
+        version = context.params.query["version"]?.try { |v| URI.unescape v }
+        tag = context.params.query["tag"]?.try { |t| URI.unescape t }
+        # Iterate over the packages to find the relevant one.
+        selected_pkg = DPPM::Prefix::Pkg.create get_prefix_or_default(from: context),
```
We don't want to create a new package, it already exists.
You can use `Prefix#new_pkg`.
I'll improve the documentation on this point.

j8r replied:
Maybe I can even delete `Pkg.create`, I'm working on that.

dscottboggs replied:
Hm. Yeah, if the package is supposed to exist already, I don't think `new` or `create` are the best choices. Perhaps `load` or something like that?

j8r replied:
The problem is the `create` method. We should only use `Prefix` to create new package objects, existing or not.
Then, we can use `#exists?` or `#exists!` to check/ensure its presence.

dscottboggs replied:
Ok excellent.

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,45 +1,137 @@
+    relative_delete "/:id/delete" do |context|
+      if context.current_user? && Config.has_access? context, Access::Delete
+        package_name = URI.unescape context.params.url["id"]
+        selected_pkg = DPPM::Prefix::Pkg.create get_prefix_or_default(from: context),
```
Same here, the package is supposed to already exist. Use `Prefix#new_pkg`.

j8r commented on src/actions/utils.cr:
```diff
@@ -0,0 +1,55 @@
+    #
+    # This macro only works inside of a route block with the variable "context"
+    # available
+    macro get_prefix_or_default(from context)
```
Use a `def`

dscottboggs commented on src/actions/pkg.cr:
```diff
@@ -1,147 +1,129 @@
+        version = context.params.query["version"]?.try { |v| URI.unescape v }
+        # Iterate over the packages to find the relevant one.
+        selected_pkg = get_prefix_or_default(from: context).new_pkg package_name, version
+        raise NoSuchPackage.new context, package_name if selected_pkg.nil?
```
Oh, `new_pkg` and `Pkg.create` don't return `Pkg?` so how will I know if the package hasn't been built?

j8r replied:
`Pkg` is package object, a virtual package if you prefer.
I may create `VirtualPkg` and `Pkg`, not done yet. For now, you can use `#exists?` and `#exists!` (`program_data.cr`)

dscottboggs commented:
The problem is that [this line](https://github.com/DFabric/dppm-rest-api/blob/39cf0c8fc1aae8d247a32b2956276bf02914a561/src/actions/pkg.cr#L44) checks for `context@current_user` before calling `has_access?` -- which we don't have an easy way to set up. I'll have to change it so that the `context.current_user?` check is in the `access_filter` proc

(this line)
```crystal
      if context.current_user? && Actions.has_access? context, Access::Delete
```

j8r commented:
I've removed `Pkg.create` in the latest version of `DPPM` - much nicer now.

j8r commented on src/actions/utils.cr:
```diff
@@ -0,0 +1,76 @@
+    # the proper boolean format, like `?value&other`
+    def parse_boolean_param(key : String, from context : HTTP::Server::Context) : Bool
+      if param = context.params.query[key]?
+        return false unless ["false", "no", "0"].includes? param.downcase
```
```suggestion
        return false unless {"false", "no", "0"}.includes? param.downcase
```
Why we also support `yes` and `no`?
Can you provide some documentation for this like MDN?

dscottboggs replied:
I can remove "no", I agree that that's silly.

dscottboggs replied:
Also I don't know where I got the idea that boolean parameters were standardized to be `?value=false` being truthy. That seems to just be wrong.

j8r commented on src/actions/app.cr:
```diff
@@ -1,185 +1,188 @@
+      if context.current_user? && Actions.has_access? context, Access::Update
+        # TODO: boot the service
+        next context
+        puts "user found!"
```
Do not forget to remove.

j8r commented on spec/spec_helper.cr:
```diff
@@ -25,13 +25,15 @@ macro assert_unauthorized(response)
+  # Disable all authentication checks by making `DppmRestApi::Actions.has_access?`
+  # always return true, then set the authentication back to the default system
+  # after yielding to the block.
+  def self.without_authentication!
```
Not a good idea to monkey patch it: by actions reading the specs, we may think it's available in the library, but it is only for testing.

j8r replied:
Not a good idea to monkey patch it: by reading the actions specs, we may think it's available in the library, but it is only for testing.

dscottboggs commented on spec/spec_helper.cr:
```diff
@@ -25,13 +25,15 @@ macro assert_unauthorized(response)
+  # Disable all authentication checks by making `DppmRestApi::Actions.has_access?`
+  # always return true, then set the authentication back to the default system
+  # after yielding to the block.
+  def self.without_authentication!
```
The property is private. I had method next to this which called this method, so that it would be at the global namespace, but I thought you wouldn't like the extra abstraction. We could also make the property not protected.

j8r commented on spec/spec_helper.cr:
```diff
@@ -25,13 +25,15 @@ macro assert_unauthorized(response)
+  # Disable all authentication checks by making `DppmRestApi::Actions.has_access?`
+  # always return true, then set the authentication back to the default system
+  # after yielding to the block.
+  def self.without_authentication!
```
I deleted my comment by mistake because I thought I have submitted it twice :/ (I said it wasn't a good idea to monkey patch be cause the method is only used in specs).

I don't understand, have you an example to show why a normal top_level method can't be used?

j8r commented on src/errors/no_such_pkg.cr:
```diff
@@ -0,0 +1,9 @@
+    setter cause
+
+    def initialize(context, package_name : String)
+      super context, "no such message found called \"" + package_name + '"'
```
```suggestion
      super context, "no such package found called \"" + package_name + '"'
```

j8r commented on src/actions.cr:
```diff
@@ -1,38 +1,21 @@
-{% end %}
-
 require "./access"
+require "./actions/utils"
```
It would be more logical to include `utils` where it is used, in each action file.

dscottboggs replied:
Utils are inherently tools that are there for all actions, so it makes more sense to me for it to be here, but if you feel strongly about it I can change it. It all goes in to the same tree anyway.

j8r replied:
If that's for all actions, it make sense.
So, I don't think the utils name is appropriate, maybe `RouteHelper`

dscottboggs replied:
Sounds good.

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,45 +1,129 @@
+        pkg = get_prefix_or_default(from: context).new_pkg package_name,
+          version: context.params.query["version"]?
+        pkg.build confirmation: false { }
+        pkg.exists!
```
I raises if the package exists, the method name is confusing. I should be `available!`
~Anyway, that's the library job to do this checks.~

j8r replied:
Currently, a package that already exists won't return an error, but it will for an application.

dscottboggs replied:
ok. Is there any indication of failure for that method? Oh... I should be raising an `HTTPStatusError` from that block, shouldn't I?

j8r replied:
What block? This method (now named `available!`) will raise an exception

dscottboggs replied:
I meant from `#build`, sorry

j8r replied:
Yes of course, there are lots of scenario for a build to fail.
The blocks is yield just before any change is done on the file-system. With the use of a boolean, we can choose to return a `400`, then after the block a `500`. What do you think?

dscottboggs replied:
> The blocks is yield just before any change is done on the file-system

Please document these inconsistent uses of block yielding. I would say that even yielding to the block on an error is unidiomatic and should be documented.

> With the use of a boolean, we can choose to return a 400, then after the block a 500.

I'm sorry I don't understand what you mean by this

j8r replied:
This is documented

dscottboggs replied:
Oh, I see... :man_facepalming:

j8r replied:
See https://github.com/DFabric/dppm/blob/master/src/prefix/pkg.cr#L109
> Build the package. Yields a block before writing on disk. When confirmation is set, the block must be true to continue.

The first stage is retrieving version, ensuring the packge exists etc.
If it fails, that's a `400 Bad Request`
Then, the block yields, and the second stage is executed, that performs changes on disk.
In any case of error, this is a `500 Internal Server Error`, because the build was supposed to work.

dscottboggs replied:
I see. :+1:

dscottboggs replied:
Ok...maybe I don't see. So... the block is yielded to if the file already exists. I get that. Then, you simulate a build, but only [`if confirmation`](https://github.com/DFabric/dppm/blob/6016bf9da7688dd02ce03557371cbd4eaacccfd4/src/prefix/program_data.cr#L86) (which `confirmation` is false in this case). When `confirmation` is false, you still yield to the block, though, just without [logging a bunch of stuff](https://github.com/DFabric/dppm/blob/6016bf9da7688dd02ce03557371cbd4eaacccfd4/src/prefix/program_data.cr#L87-L103).

It looks like,  as a library user, I have to *count* the number of times the block is called, and determine which error to throw based on said count?

I realize it's getting quite late in your timezone so we can put this off until tomorrow if you prefer.

j8r replied:
The block is always yielded, and only once.
So we can have:
```cr
build_init = false
pkg.build confirmation: false { build_init = true }
```

j8r replied:
We should also support simulate, later. The DPPM library isn't ready for that yet.

dscottboggs replied:
This feature doesn't seem to work according to my test? Please see [the latest CI results](https://cloud.drone.io/DFabric/dppm-rest-api/214/1/5) when you get a chance.

dscottboggs commented on spec/spec_helper.cr:
```diff
@@ -25,13 +25,15 @@ macro assert_unauthorized(response)
+  # Disable all authentication checks by making `DppmRestApi::Actions.has_access?`
+  # always return true, then set the authentication back to the default system
+  # after yielding to the block.
+  def self.without_authentication!
```
in the commit I just made -- `protected method 'access_filter=' called for DppmRestApi::Actions:Module`

dscottboggs commented on spec/spec_helper.cr:
```diff
@@ -43,6 +41,12 @@ Fixtures.reset_config
-Spec.before_each { Fixtures.reset_config }
+Spec.before_each do
+  Fixtures.reset_config
+  FileUtils.mkdir_p Fixtures::PREFIX_PATH
```
Is there a way for DPPM to instantiate the directory structure that it needs? I'm getting a *no such file or directory* error when trying to clean packages on an empty prefix.

j8r replied:
Yes, `Prefix#create`

j8r commented:
Do not forget to `shards update` for the get latest DPPM commit :wink:

j8r commented on spec/fixtures.cr:
```diff
@@ -12,6 +12,8 @@ module Fixtures
     NORMAL_USER = Random::Secure.base64(TEST_KEY_SIZE)
   end
 
+  PREFIX_PATH = Path[__DIR__, "prefix"].to_s
```
```suggestion
  PREFIX_PATH = File.tempname("_temp_dppm_prefix")
```

j8r commented on src/actions/utils.cr:
```diff
@@ -0,0 +1,79 @@
+      end
+    end
+
+    def ensure_dir_exists(dir : String)
```
This method shouldn't exist. All file_system operations should stay in the DPPM library.
Furthermore, the DPPM prefix should be already created when running the webserver, no need to check it anymore.

j8r reviewed (commented):
IMHO the prefix doesn't have to be configurable.
It would be safer, and also I don't see a use case since we have namespaces.

dscottboggs committed:
Update spec/fixtures.cr

Co-Authored-By: Julien Reichardt <mi@jrei.ch>

dscottboggs commented:
> IMHO the prefix doesn't have to be configurable.

That's a fair point. However, that's a very major change and would involve updates to the CLI and `Config` types, so I think it would be easier to make that a separate PR and continue with this as though that weren't going to happen.

j8r commented on src/actions/utils.cr:
```diff
@@ -0,0 +1,79 @@
+    # This macro only works inside of a route block with the variable "context"
+    # available
+    def get_prefix_or_default(from context)
+      DPPM::Prefix.new ensure_dir_exists context.params.query["prefix"]? || DPPM.default_prefix
```
You can use `Prefix#create`, that ensures the directories exist. `ensure_dir_exists` can then be removed.
As you said, this is temporary.

dscottboggs committed:
Further work on testing

Update DPPM namespacing

Co-Authored-By: Julien Reichardt <mi@jrei.ch>

Successfully test build route

Update dependencies

Add specs for config querying

j8r commented on spec/spec_helper.cr:
```diff
@@ -25,13 +25,21 @@ macro assert_unauthorized(response)
+  # Disable all authentication checks by making `DppmRestApi::Actions.has_access?`
+  # always return true, then set the authentication back to the default system
+  # after yielding to the block.
+  def self.without_authentication!
```
After thinking more, I think it's would be fine to monkey patch if we explicitly state this is used for specs:
```cr
def DppmRestApi.spec_without_authentication!(&block)
  Actions.access_filter = ->(_context : HTTP::Server::Context, _permissions : Access) { true }
  yield
  Actions.access_filter = ->access_filter(HTTP::Server::Context, Access)
end
```

dscottboggs replied:
I don't like the phrasing of that. What if we create a `SpecHelpers` module (like we have with `RouteHelpers`), so that it's `SpecHelpers.without_authentication!`?

j8r replied:
That's also good, as long as the name "spec" or "test" is inside to show that it is only used in spec tests.

j8r commented on spec/spec_helper.cr:
```diff
@@ -25,13 +25,21 @@ macro assert_unauthorized(response)
-  DppmRestApi::Actions.access_filter = DppmRestApi.access_filter
+  DppmRestApi.without_authentication! { yield }
 end
 
```
```suggestion
DPPM::Log.output = DPPM::Log.error = File.open File::NULL, "w"
```
I've move `Log` to the `DPPM::Log` namespace in a new commit

j8r commented:
I don't think making the prefix not configurable means to change evertyhing.
Just make sure `DppmRestApi.run` take also a prefix as argument, then create a protected `class_property`  in `Actions`.

I think it's ok to have the prefix configurable in the CLI, mainly because  the`dppm` command can be available globally on the system, but not when running the web rest api server.
By default, the dppm application will manage its prefix where it is installed – nice!

j8r commented:
I don't think making the prefix not configurable means to change everything.
Just make sure `DppmRestApi.run` take also a prefix as argument, then create a protected `class_property`  in `Actions`.

I think it's ok to have the prefix configurable in the CLI, mainly because  the`dppm` command can be available globally on the system, but not when running the web rest api server.
By default, the dppm application will manage its prefix where it is installed – nice!

dscottboggs committed:
Refactor prefix API (again)

Remove the ability to configure the prefix path via query parameter

Don't automatically create a prefix

Co-Authored-By: Julien Reichardt <mi@jrei.ch>

j8r commented on src/dppm_rest_api.cr:
```diff
@@ -28,12 +28,14 @@ module DppmRestApi
       @@permissions_config = Config.from_json data
     end
 
+    DppmRestApi.permissions_config.prefix = DPPM::Prefix.new(prefix).tap &.create
```
```suggestion
    DppmRestApi.permissions_config.prefix = DPPM::Prefix.new prefix
```
Bad idea, I don't think we should create prefixes automagically.
The user have to explicitly do `dppm app install`. This can create a system service. This will be a way to start the API.
To check if the Prefix exists, use `Prefix#check`

j8r commented on src/config.cr:
```diff
@@ -1,12 +1,25 @@
+  def prefix=(prefix : DPPM::Prefix)
+    @prefix = prefix.path.to_s
+    @_prefix = prefix
+  end
```
All of this isn't needed. Just `class_property : DPPM::Prefix { raise "no prefix set" }`.
The prefix don't have to be in the configuration, the sane default prefix managed should be where the `dppm` application is.

j8r commented on src/cli.cr:
```diff
@@ -13,6 +13,10 @@ module DppmRestApi::CLI
+        prefix: {
+          info: "The base directory to install dppm packages into",
+          default: DPPM.default_prefix
+        }
```
No, it's already inherited.

dscottboggs replied:
oh ok

j8r commented on src/dppm_rest_api.cr:
```diff
@@ -11,7 +11,9 @@ module DppmRestApi
   end
 
-  private def self.access_filter(context : HTTP::Server::Context, permission : Access) : Bool
+  class_property prefix : DPPM::Prefix { raise "no prefix set" }
```
That's a good idea to put it there :+1:

j8r replied:
I think it will be better in `Actions.prefix`, that were the prefix is used.

j8r commented on spec/permissions.json:
```diff
@@ -0,0 +1,74 @@
+      "name": "Jim Oliver"
+    }
+  ]
+}
```
?

scott-bossanova replied:
wtf...?

dscottboggs replied:
There is an exception being thrown when the tests are run, this file doesn't get cleaned up because of that exception

j8r replied:
Why it is written there? It should be in the temp directory

dscottboggs replied:
That's a very good question....

j8r replied:
That's because of `Fixtures::PERMISSION_FILE`, it should use `File.tempname("_temp_dppm_permisions_file")`

dscottboggs replied:
No, because the filename "permissions.json" is hardcoded, the server CLI only gets the directory that it's supposed to be stored in. In the case of the specs, that's `__DIR__`, which resolves to `$PROJECT_DIR/spec`.

j8r replied:
Not a problem, all the `data_dir: __DIR__` in `cli_spec` can be changed to the `Fixtures` path.

dscottboggs replied:
So... perhaps, instead of `__DIR__` and `Fixtures::PERMISSION_FILE`, we could have `Fixtures::DIR`? Furthermore, should this directory be within the project directory or within the system's tempdir?

j8r commented:
I suggest to move the `class_property prefix` to `Actions`. It's only used there AFAIK (the CLI will still have a configurable prefix).
The module will be more independent, too.

dscottboggs committed:
Additional testing; Simplify build flow

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,45 +1,142 @@
+        init_done = false
+        begin
+          pkg.build confirmation: false { init_done = true }
+        rescue e
```
Nice! Little pedantic suggestion:
```cr
pkg.build confirmation: false do
  init_done = true
rescue e
...
end
```

dscottboggs replied:
This still gives an *Internal Server Error* instead of *Bad Request* when given a package name that does not exist. In fact, I don't see any scenario in which an exception will be raised before [the block is yielded](https://github.com/DFabric/dppm/blob/fa4faae67e475f87c820020854ce71c6309f6074/src/prefix/pkg.cr#L143-L150).

j8r replied:
Good point, sorry

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,45 +1,142 @@
+      if Actions.has_access? context, Access::Read
+        pkgs = [] of String
+        begin
+          DppmRestApi.prefix.each_pkg { |pkg| pkgs << pkg.package }
```
It shouls be a list of hashes, with at least the package name and the version. I think it's a sane default.
In the future (or now if you prefer), this elements will be configurable through parameters.
An user may wants also descriptions, documentation link, license etc.

dscottboggs replied:
> this elements will be configurable through parameters. An user may wants also descriptions, documentation link, license etc.

A user will not be using the API directly. If the client wants to query that information about a package, they could use the *query* route.

j8r replied:
Why? This means when the Web client will list all applications, it will also have to do multiple requests?
We will have duplicate packages with this current implementation when we have two versions of the same package.

dscottboggs replied:
Fair enough.

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,45 +1,142 @@
+          DppmRestApi.prefix.each_pkg { |pkg| pkgs << pkg.package }
+        rescue e
+          raise InternalServerError.new context,
+            message: "#each_pkg raised an exception with the message '#{e.message}'.",
```
This is an internal implementation detail, perhaps something more generic like "Package listing failed".
No need to add the exception message inside it, it will be shown already in the error trace.

j8r commented on src/actions/route_helpers.cr:
```diff
@@ -0,0 +1,61 @@
+    # or `"0"`.
+    def parse_boolean_param(key : String, from context : HTTP::Server::Context) : Bool
+      if param = context.params.query[key]?
+        return true unless {"false", "0"}.includes? param.downcase
```
Any reason I miss to support also `0`, in addition to `false`?

dscottboggs replied:
Idk it just seems common to assume 1 and 0 will work for boolean operations.

j8r replied:
Not really, we use `true` and `false` everywhere else (json, con, CLI, Crystal).
It seems also more common for other APIs to pass `true` and `false` (which are strings)

j8r replied:
The `param.downcase` means  that `FALSE` is also supported?!

j8r replied:
I propose to only support `true` and `false`, else raise.

dscottboggs replied:
I hadn't thought of raising `BadRequest`. Counter-proposal -- only support the existence or lack thereof and raise if there is *any* text given in a boolean parameter. `?libraries&something` would be the only option, anything else ( `?libaries=yes&something=false`) would raise `BadRequest`.

j8r replied:
Wait, why having this possibility?
I think I will rename `query` to `info`, because that's the info present in `pkg.con` file.
Querying the configuration is another endpoint, `/config`, and abstract the origin of the configuration.
The endpoint could be `/pkginfo`

dscottboggs replied:
Is this comment about a different section of code? I'm confused

j8r reviewed (commented):
Good overall! Some notes on advanced features.

dscottboggs committed:
Change format of list pkgs response

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,45 +1,155 @@
+        pkgs = {} of String => ListResponse
+        DppmRestApi.prefix.each_pkg do |package|
+          if (found = pkgs[package.package]?) && found.version > package.semantic_version
+            # newer version of package already exists
```
Why does it matter? We want to list **all** packages.
I think the best would be an array of hash of string => string.

dscottboggs replied:
OK so the structure of the response is like so

```json
{ "data": {
  "dppm": {
    "version": "2019.4.26"
  }
} }
```

This leaves room for expanding the API to have further information such as 

```json
{ "data": {
  "dppm": {
    "version": "2019.4.26",
    "license": "ISC",
    "description": "Dedicated Platform Package Manager - an easy way to install and manage server applications."
  }
} }
```
I thought you wanted me to remove duplicates in this method but now that I think about it that doesn't really make sense. I'll get rid of that part.

j8r replied:
What duplicates? With the examples you posted, it can't have several `dppm` packages, which can happen when upgrading for example.

j8r replied:
I propose:
```json
"data":[
{
  "package": "dppm",
  "version": "1.2.3"
},
{
  "package": "dppm",
  "version": "1.2.7"
}
]
```
Of course it can also include description, license etc. as you said, depending of parameters.
It would be more or less a list of what's in `PkgFile`, plus the version.

dscottboggs committed:
Change base dir for tests from __DIR__ to a File.tempname in Fixtures

dscottboggs commented on spec/actions/pkg_spec.cr:
```diff
@@ -165,10 +165,10 @@ module DppmRestApi::Actions::Pkg
       end
     end
-    it "responds with Internal Server Error when given an invalid package name" do
+    it "responds with Bad Request when given an invalid package name" do
```
This spec is failing due to an issue in the library I think.

j8r replied:
The dppm library? I don't think so, it doesn't know the concept of API routes. I figured out why.

dscottboggs replied:
The dppm library doesn't need to know API routes to raise an exception upon receiving an invalid package name. Which is what this spec is testing. Sorry, I was trying to point out [the body](https://github.com/DFabric/dppm-rest-api/blob/cfe755532ecbfb529208268df793ccd2b7482ef4/spec/actions/pkg_spec.cr#L170-L172), not which HTTP status was raised..

dscottboggs replied:
```crystal
post fmt_route "/$(echo all your base are belong to us!)/build"
# (which does)
pkg = pfx.new_pkg "$(echo all your base are belong to us!)", version: nil
# this fails both because there's no underscore in the package name
# with no version specified, and because the package name has invalid
# characters
init_done = false
pkg.build confirmation: false do
  # this block should never get yielded to, due to input validations raising
  # an exception before the block is called
  init_done = true
rescue e
  # which means this exception is raised.
  raise InternalServerError.new context, cause: e if init_done
  raise BadRequest.new context, cause: e
end
# making this line fail
response.status_code.should eq HTTP::Status::BAD_REQUEST.value
```

j8r replied:
This is due to some methods not being catched in `Actions::Pkg` (see my code suggestion).
A package can be built internally from a source (`Src` object) in order to take the latest version. If it doesn't exist, an exception is raised.
Setting the package name + version doesn't require to read a source package.

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,45 +1,152 @@
+
+    struct ListResponse
+      include JSON::Serializable
+      property version : SemanticVersion
```
All code relative to `SemanticVersion` isn't needed.
Also, what the `ListResponse` abstraction bring?
We will need similiar objects too for `app` and `src`.
The difference between the three will be:
- src: nothing, what's in pkgfile.
- pkg: add version
- app: add name (and perhaps version)

dscottboggs replied:
It's not an abstraction, it's a type. We could use a NamedTuple if you're against having a defined struct...tbh I'm kind of confused as to what the issue is. The app and src versions can be `Actions::App::ListResponse` and `Actions::Src::ListResponse`.

j8r replied:
Does JSON.build directly wouldn't be better? In the sense of being flexible for what fields we return to the client, like description, info, etc.
This will avoid having 3 types, and the logic can be mostly shared.

dscottboggs replied:
But why? What makes JSON.build better? It would be more code, and more prone to mistakes.

j8r replied:
How do you do to build only fields requested?

dscottboggs replied:
Oh. That's a good point. If you want to do that, then yes, a `build` implementation would be more appropriate. If we are allowing bulk querying of config values, then I propose renaming this method to `/dppmrestapi/pkg/query`, and offer optional pagination on the response as a long-term goal once the project is more developed.

j8r replied:
... that's a good idea, I may implement this in the CLI at first.
Ok, we could let this as is and add a `TODO` comment.

dscottboggs replied:
What config values were you thinking of allowing in the parameters for this? I didn't see anything relevant in `Pkg` or `Base`, unless you were planning to allow querying arbitrary config values in this route?

j8r replied:
Mainly what's in `PkgFille`, plus version for package and name for app.
This will help when we will build the Web client, listing all apps will require only one request.

j8r commented on src/ext/semantic_version.cr:
```diff
@@ -0,0 +1,27 @@
+require "semantic_version"
```
This file isn't needed.

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,45 +1,152 @@
+          init_done = true
+        rescue e
+          raise InternalServerError.new context, cause: e if init_done
+          raise BadRequest.new context, cause: e
```
This should be:
```cr
begin
  pkg = pfx.new_pkg package_name, version: context.params.query["version"]?
  pkg.build confirmation: false { init_done = true }
rescue ex
  raise InternalServerError.new context, cause: ex if init_done
  raise BadRequest.new context, cause: ex
end
```

j8r commented on spec/spec_helper.cr:
```diff
@@ -25,21 +25,33 @@ macro assert_unauthorized(response)
 
+DPPM::Log.output = DPPM::Log.error = File.open File::NULL, "w"
 Kemal.config.env = "test"
+FileUtils.mkdir_p Fixtures::DIR.to_s
```
```suggestion
FileUtils.mkdir_p Fixtures::DIR.to_s
at_exit { FileUtils.rm_rf Fixtures::DIR.to_s }
```
Add also `require "file_utils"`

j8r commented on spec/spec_helper.cr:
```diff
@@ -25,21 +25,33 @@ macro assert_unauthorized(response)
 # Set all configs back to the expected values, in case they changed
-Spec.before_each { Fixtures.reset_config }
+Spec.before_each do
+  FileUtils.mkdir_p Fixtures::DIR.to_s
```
Already present
```suggestion

```

j8r commented:
Running spec is slow because each operation involve downloading `packages-source` archive and the `dppm` binary,  again and again.
I propose mocking instead.
Look at `./lib/dppm/spec/samples/testapp`
Edit: I can fix this in another PR, I guess.

j8r committed:
Use mock data

j8r committed:
Move DppmRestApi.prefix to Actions.prefix

j8r commented:
I have taken the initiative to push some commits. Be free to put comments on points you disagree.

dscottboggs commented on spec/actions/pkg_spec.cr:
```diff
@@ -53,8 +53,7 @@ module DppmRestApi::Actions::Pkg
   def build_test_package
-    DppmRestApi.prefix.update
-    pkg = DppmRestApi.prefix.new_pkg "dppm"
+    pkg = DppmRestApi.prefix.new_pkg "testapp"
```
Why does this work?

dscottboggs commented on spec/spec_helper.cr:
```diff
@@ -49,6 +50,7 @@ Spec.before_each do
   Fixtures.reset_config
   FileUtils.mkdir_p Fixtures::PREFIX_PATH.to_s
   DppmRestApi.prefix.create
+  FileUtils.cp_r "./lib/dppm/spec/samples", DppmRestApi.prefix.src.to_s
```
Oh, I see!! :100:

dscottboggs commented:
> I have taken the initiative to push some commits. Be free to put comments on points you disagree.

I was working on this process locally and this seems to fix all the issues I had and more. Thanks for your help! I'll get to some other tasks now.

dscottboggs committed:
Use JSON.build for multi-pkg query response

dscottboggs committed:
Improve error handling

j8r commented on src/actions/route_helpers.cr:
```diff
@@ -15,6 +15,11 @@ module DppmRestApi::Actions
+        # Catch non-HTTPStatusError exceptions, and throw an InternalServerError
+        # with the original error as the cause.
+        raise e if e.is_a? HTTPStatusError
+        raise InternalServerError.new \{{block.args[0]}}, cause: e
```
I think it would be clearer to do something like:
```cr
...
rescue ex : Actions
  raise ex
rescue ex
  InternalServerError
end
```
This means moving exception dedicated to `Actions`, in the `Actions` module.

dscottboggs replied:
Are you suggesting that [this](https://github.com/DFabric/dppm-rest-api/blob/cfe755532ecbfb529208268df793ccd2b7482ef4/src/errors/http_status_errors.cr#L5) be renamed to `Actions` or `ActionException` or something? Or perhaps an alias like `alias Actions = HTTPStatusError | SomethingElse`?

j8r replied:
I submitted a commit to show what I was thinking: having `Actions::Exception`.

j8r committed:
Move custom exceptions to Actions::Exception

j8r committed:
Change relative_* macros to methods

dscottboggs commented on src/actions/route_helpers.cr:
```diff
@@ -1,68 +1,67 @@
+
+  {% for method in %w(get post put delete ws options) %}
+  # Perform a relative {{method.id}} from the `ROOT_PATH`.
+  def relative_{{method.id}}(route : String?, &block : HTTP::Server::Context ->)
```
Why??

j8r replied:
Because better to avoid unneeded macros. A classic method do the job.

dscottboggs replied:
> better to avoid unneeded macros

What makes you say that? I don't see anything wrong with making a liberal use of macros. That said, after giving it some thought, in this case it does neaten things up a bit.

j8r replied:
I disagree, macros have strong downsides. That why they must not be abused, and **only** used when strongly needed.
- harder to debug, specially compile time errors
- slower compile times
- lazy typing, no type for arguments, no signature
- unwanted side effects (modifying variables in the current scope)

dscottboggs committed:
More explicit and strict handling of boolean parameters

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,44 +1,150 @@
+                json.object do
+                  json.field "package", package.package
+                  json.field "version", package.version
+                  package.each_config_key do |key|
```
I don't really see the point for packages. However for applications it will be useful.

dscottboggs replied:
so leave this off, and we're ready to merge?

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,44 +1,150 @@
+                  package.each_config_key do |key|
+                    if parse_boolean_param key, from: context
+                      json.field key do
+                        package.get_config(key) { nil }.to_json json
```
Use `get_config` instead, which will raise.
Note there is also `get_config?`

dscottboggs replied:
> Use get_config instead, which will raise.

Note that the current implementation of `build_json` requires that [no exceptions be thrown](https://github.com/DFabric/dppm-rest-api/blob/adbae12b5e71c925ff0a1404631902849a4c3385/src/actions/route_helpers.cr#L42-L62) while the response is being built. Until now, the only resolution to this I could think of was building to an intermediary IO and copying to `context.response` after it's built. That would be a big performance hit in a crucial spot. However, talking another look at my implementation I was thinking that there is one alternative, although it is less than ideal....suggestions are welcome, but I'd recommend we discuss this in a separate PR, so that I can also begin work on other routes while we work out the details.

```crystal
def build_json(response : IO)
  error : Exception? = nil
  JSON.build response do |json|
    json.object do
      json.field "data" do
        json.object do
          yield json
        rescue e
          error = e
        end
      end
      if error
        json.field "errors" do
          ErrorResponse.new(error).errors.to_json json
        end
      end
    end
  end
end
```

I realize this would require specifying a partial succesful response, but there's no way to rewind an `HTTP::Server::Response` because it's *streamed* out to the client on the fly.

dscottboggs replied:
For the package routes, I (think) I successfully avoided cases which raise, so that's an option as well, but it's...fragile, and I don't like it, especially not long-term

j8r replied:
Yes, we will have to figure out how to do it. We could have an abstraction for this and use it anywhere we need.
For a reason or another, building the JSON may result of an error.

j8r commented on src/actions/route_helpers.cr:
```diff
@@ -0,0 +1,70 @@
+    context.params.query[key]?.try do |value|
+      raise BooleanParamHasValue.new context, key, value unless value == ""
+      return true
+    end
```
I propose something a bit more idiomatic:
```cr
if value = context.params.query[key]?
  return true if key.empty?
  raise BooleanParamHasValue.new context, key, value
end
false
```

dscottboggs committed:
rephrase #parse_boolean_param

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,44 +1,143 @@
+      # was found and we can query it
+      if key = context.params.query["get"]?
+        data = begin
+          selected_pkg.get_config key
```
```suggestion
         config_key = selected_pkg.get_config key
```
Then remove `data = begin`
